File: bst.py
# Generic: yes
# Type checked: no

class Node():
	def __init__(self, value):
		self.value = value
		self.l = None
		self.r = None
	def contains(self, value):
		if self.value == value:
			return True
		if value < self.value:
			if self.l == None:
				return False
			return self.l.contains(value)
		if value > self.value:
			if self.r == None:
				return False
			return self.r.contains(value)
	def add(self, value):
		if self.value == value:
			return
		if value < self.value:
			if self.l == None:
				self.l = Node(value)
			else:
				self.l.add(value)
		if value > self.value:
			if self.r == None:
				self.r = Node(value)
			else:
				self.r.add(value)


root = Node(10)
root.add(20)
root.add(15)
root.add(20)

for n in [10,20,15,1]:
	print str(n) + ":" + str(root.contains(n))

File: bst.cpp
// Generic: yes
// Type checked: yes

#include <iostream>

template <class T>
class Node
{
private:
	Node<T> *l, *r;
public:
	T value;
	Node(T value):l(nullptr), r(nullptr), value(value){
		this->value = value;
	}
	~Node(){
		if(this->l) delete this->l;
		if(this->r) delete this->r;
	}
	bool contains(T value){
		if(value == this->value) return true;
		else if (value < this->value){
			if(this->l) return this->l->contains(value);
			else return false;
		}
		else {
			if(this->r) return this->r->contains(value);
			else return false;
		}
	}
	void add(T value){
		if(value == this->value) return;
		else if (value < this->value){
			if(this->l) this->l->add(value);
			else this->l = new Node<T>(value);
		}
		else if (value > this->value){
			if(this->r) this->r->add(value);
			else this->r = new Node<T>(value);
		}
	}
};

int main(int argc, char** argv){
	Node<int> root(10);
	root.add(20);
	root.add(15);
	root.add(20);
	std::cout << root.contains(10) << root.contains(15) << root.contains(20) << root.contains(1) << "\n";
	return 0;
}